wallet/listeners/transaction_listener.go

```diff
@@ -212,6 +219,18 @@ func (l *TransactionListener) processSalePayment(txid string, output wallet.Tran
 	l.db.TxMetadata().Put(repo.Metadata{txid, "", title, orderId, thumbnail, bumpable})
 }
 
+func currencyDivisibilityFromContract(mw multiwallet.MultiWallet, contract *pb.RicardianContract, orderID string) uint32 {
+	var currencyDivisibility = contract.VendorListings[0].Metadata.CoinDivisibility
+	if currencyDivisibility != 0 {
+		return currencyDivisibility
+	}
+	wallet, err := mw.WalletForCurrencyCode(contract.BuyerOrder.Payment.Coin)
+	if err == nil {
+		return uint32(wallet.ExchangeRates().UnitsPerCoin())
+	}
+	return core.DefaultCurrencyDivisibility
```

Comment:
Prefer divisibility from listing, then check wallet's preference, then fail to default as a last resort.